src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -994,13 +930,10 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
 
         FoundNonAsciiDataInFirstVector:
```

Comment:
Yes, this would compute `MoveMask` twice for SSE2 when a string contains non-ASCII if JIT couldn't spot it.
I didn't test this but JIT may be able to avoid this duplication as the mask computation is right after the non-ASCII check. For constants, JIT reused the value available in reg, e.g., `xmm1` is loaded only once [here](https://gist.github.com/SwapnilGaikwad/e4e65798b924567c57cd0588a5c4740a#file-less_args_no_debug_asm_dump-txt-L141).

This simplifies refactoring and adding the AArch64 version as part of the issue #41292. What would you suggest?

Reply:
You should verify it, but I wouldn't change the existing unrelated methods that has chances of degrading the perf.

Reply:
[Assembly](https://gist.github.com/SwapnilGaikwad/e0132037e6500af9d8d19f3bebcb6e14#file-sse2_only_vectorcontainsnonasciichar-txt-L83-L85) shows that the duplicate computation of `moveMask` cannot be avoided.
I agree that we shouldn't make existing things slower but still think that we should take this penalty because:
- It's a pentaly of a few cycles in the error case when a non-ASCII char is found in a given chunk on the systems pre-SSE4.1
- It will make it easier to move to the vector API when we can implement `VectorContainsNonAsciiChar` purely using the API
- This patch adds similar pentaly of a few cycles on AArch64 while detecting presence of non-ASCII char in the first or second chunk. [There](https://gist.github.com/SwapnilGaikwad/f5533c5e57e0a277ef26305e716c730d#file-aarch64_less_args_no_debug_asm_dump-txt-L119-L120) we repeate pairwise-max and SIMD to GPR transfer for unified AArch64 and x86 version. We would need more platform specific changes to the loop for reclaiming those cycles, e.g. [here](https://github.com/dotnet/runtime/pull/70080/commits/6ac5e90a8142faef700efbab714250db7ac443ff#diff-66bbe89271f826c9232bd146abb678844754515dc027f70ad0ce36f751da46ebR1661-R1667).

If you want to avoid the additional mask computation then I can [undo refactoring](https://gist.github.com/SwapnilGaikwad/932ff832b342f9c4d9437d71ec21a62b).

Reply:
I think it's fine and we can log an issue to track improving the codegen.

Reply:
Agreed with Kunal. I'll keep it as is then.